Change Switch Network modal for a dropdown with only the target network

tokodev opened the pull request:
As commented here: [issue 300](https://github.com/scaffold-eth/se-2/issues/300#issuecomment-1512725914) to improve the user experience and simplify the network switching (which had networks that may confuse the user) I've replaced the existing modal with a dropdown menu that appears when the user is on the wrong network. 

The dropdown menu only includes the target network switch option and the disconnect option. Here is a video of the new behaviour: [PR se-2 issue-300 - Watch Video](https://www.loom.com/share/82e23277107548479d26afbd3c019eea)

carletex commented:
Looking good @tokodev thanks! Like I said in #300 I think it makes more sense to have it like this. A modal for just a couple of items is less friendly.

Assigning @technophile-04 to test & review.

technophile-04 reviewed (commented):
Tysm @tokodev looks nice. Tested it out and it works similarly to `main` branch 

Added a few small comments 🙌

technophile-04 commented on packages/nextjs/components/scaffold-eth/RainbowKitCustomConnectButton.tsx:
```diff
@@ -31,15 +34,43 @@ export const RainbowKitCustomConnectButton = () => {
-                  </>
+                    <ul tabIndex={0} className="dropdown-content menu p-2 shadow bg-base-100 rounded-box w-56">
+                      <li>
+                        <a
```
We should use `button` here instead of `a` because it changes the URL to `/#` due to href every time users click it

tokodev replied:
Thx for the review! 🙌 I've just pushed a commit with those tweaks:
- Changed both `a` for `button`
- Removed comments and `className="text-red-500"` from the childs, leaving it in parent element

tokodev replied:
I should have taken a look on the code and see there was already a `button` for this purpose instead of an `a`, sorry 😅
Fixed now!

technophile-04 commented on packages/nextjs/components/scaffold-eth/RainbowKitCustomConnectButton.tsx:
```diff
@@ -31,15 +34,43 @@ export const RainbowKitCustomConnectButton = () => {
+                        >
+                          <ArrowLeftOnRectangleIcon className="h-6 w-4 ml-2 sm:ml-0 text-red-500" />{" "}
+                          {/* Add the 'text-red-500' class */}
+                          <span className="text-red-500">Disconnect</span> {/* Add the 'text-red-500' class */}
```
Remove the comment, also we can just add `className="text-red-500"` to parent element and it will cascade its effects to child's so you don't have to mention `className="text-red-500"` on each child

tokodev replied:
Done!

tokodev commented on packages/nextjs/components/scaffold-eth/RainbowKitCustomConnectButton.tsx:
```diff
@@ -31,15 +34,43 @@ export const RainbowKitCustomConnectButton = () => {
+                      </li>
+                      <li>
+                        <a
+                          className="menu-item text-red-500" // Add the 'text-red-500' class
```
Done!

tokodev commented:
Thx for the review @technophile-04 ! 🙌 I've just pushed a commit with those tweaks:

Changed both a for button
Removed comments and className="text-red-500" from the childs, leaving it in parent element

technophile-04 commented:
Tysm @tokodev , everything works nicely 🙌

One small bug which I noticed while testing (its present in `main` branch too) : 
If you switch to wrong network second time when you are currently on wrong network it disconnects your wallet 

https://user-images.githubusercontent.com/80153681/233035324-775d1357-c1b4-460f-9c89-b73e4f5daaaa.mov

 

cc @carletex for final review and merge

sverps commented on packages/nextjs/components/scaffold-eth/RainbowKitCustomConnectButton.tsx:
```diff
@@ -31,15 +34,41 @@ export const RainbowKitCustomConnectButton = () => {
+                          }}
+                        >
+                          <ArrowsRightLeftIcon className="h-6 w-4 ml-2 sm:ml-0" />
+                          <span>
```
Add `className="whitespace-nowrap"` and then on line 42, remove `w-56` to to let the dropdown set its width based on the text (allows it to vary based on what the target network name is)

tokodev replied:
Thanks! Cool to know about `whitespace-nowrap` utility

carletex commented on packages/nextjs/components/scaffold-eth/RainbowKitCustomConnectButton.tsx:
```diff
@@ -31,15 +34,41 @@ export const RainbowKitCustomConnectButton = () => {
+                        <button
+                          className="menu-item text-red-500"
+                          type="button"
+                          onClick={() => {
+                            disconnect();
+                          }}
```
Interested in knowing why this happens XD @sverps 

Never seen it before, for me they should be the same thing and I'd always use `onClick={disconnect}`

tokodev replied:
Updated with `onClick={() => disconnect()}`

technophile-04 replied:
Here is the parsed error : 
![Screenshot 2023-04-20 at 3 31 18 PM](https://user-images.githubusercontent.com/80153681/233333185-2b21a005-308e-4318-9c42-dfb248056085.jpg)

The type of `disconnect` : `UseMutateFunction<void, Error, void, unknown>`  
typeof `onClick` for button : `MouseEventHandler<HTMLButtonElement>`

The place where their both types mismatch in function `parameters` : 

Actual type of `onClick` is somewhat : `(arg : MouseEvent<HTMLButtonElement, globalThis.MouseEvent>) => ...` 

and I think resolved type of `disconnect` is: `(arg : void) => ...`

so the `arg` is mismatched, 

--- 

Doing `onClick={() => discconect()}` works because I think here we have just inlined function without assigning any types to inlined function not even to  `args`  and its all infered 

If we do `onClick={(e: void) => disconnect()}` we get same error as above : 
![Screenshot 2023-04-20 at 3 46 05 PM](https://user-images.githubusercontent.com/80153681/233336699-42af63e2-4486-462c-b215-d36cfc885cac.jpg)


Lol that's my two cents of knowledge 😂...but would really love to hear from @sverps about the error and also please free to point out the wrong thing it will help fill in my knowledge gaps 🙏

sverps replied:
Yeah, your explanation is indeed correct.

The `disconnect` function comes from tanstack react-query and it's a mutate function, which takes the variables of the mutation as a first param.
```typescript
const disconnect: (variables: void, options?: MutateOptions<void, Error, void, unknown> | undefined) => void
```

In this case, it's a mutation without variables, hence the `void`.

If we were to place it directly in the `onClick`, it would try to pass down the `MouseEvent` as a first param, which would indeed be wrong. So typescript is there again to save the day :sweat_smile:

carletex replied:
Thanks guys <3

tokodev commented on packages/nextjs/components/scaffold-eth/RainbowKitCustomConnectButton.tsx:
```diff
@@ -31,15 +34,41 @@ export const RainbowKitCustomConnectButton = () => {
+                          className="menu-item"
+                          type="button"
+                          onClick={() => {
+                            if (switchNetwork) {
+                              switchNetwork(configuredNetwork.id);
+                            }
```
Simplified ☺ ty!

tokodev commented:
I've just pushed a commit with the tweaks after @sverps review. Tysm ❤

carletex commented:
Looking great @tokodev!! Added a tiny tweak for a bigger shadow and a bit of margin.

Thanks all for work & review <3

carletex commented:
Also changed `text-red-500` to our daisy theme red/error color.
